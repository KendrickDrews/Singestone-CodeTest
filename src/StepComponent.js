import React from 'react';
export default class StepComponent extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            error: null,
            isLoaded: false,
            items: []
        };
    }
    componentDidMount() {
        fetch("https://uqnzta2geb.execute-api.us-east-1.amazonaws.com/default/FrontEndCodeChallenge")
            .then(res => res.json())
            .then((result) => {
                this.setState({
                    isLoaded: true,
                    items: result
                });
            },                
                (error) => {
                    this.setState({
                        isLoaded: true,
                        error
                    });
                });
    }
    render() {
        const { error, isLoaded, items } = this.state;
        const finalArray = items.sort(function (a, b) {
            return a.stepNumber - b.stepNumber;
        }).map(item => {
            const finalItem = {};
            const versionDates = item.versionContent;
            const sortByEffectiveDate = versionDates.sort(function (a, b) {
                return Date.parse(b.effectiveDate) - Date.parse(a.effectiveDate);
            });
            finalItem.stepNumber = item.stepNumber;
            finalItem.title = sortByEffectiveDate[0].title;
            finalItem.body = sortByEffectiveDate[0].body;
            return finalItem;
        });
        if (error) {
            return <div>Error: {error.message}</div>;
        }
        else if (!isLoaded) {
            return <div>Loading...</div>;
        }
        else {
            return (<div className="step-row">
                {finalArray.map((item) => (<div className="step-container" key={item.stepNumber}>
                    <p className="step-stepNumber">0{item.stepNumber}</p>
                    <hr className="step-separator"></hr>
                    <p className="step-title">{item.title.toUpperCase()}</p>
                    <p className="step-body">{item.body}</p>
                </div>))}
            </div>);
        }
    }
}
