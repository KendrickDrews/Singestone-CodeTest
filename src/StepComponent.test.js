import React from 'react';
import Enzyme from 'enzyme';
import StepComponent from './StepComponent';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
 
Enzyme.configure({ adapter: new Adapter() });

it('renders loading page on init', () => {
    let wrapper = Enzyme.mount(<StepComponent/>);   
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders error page on error', () => {
    let wrapper = Enzyme.mount(<StepComponent/>);   
    wrapper.setState({
        error: {message:'test error'},
        isLoaded: true,
        items: []
    });
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders steps with mock step data', () => {
    let wrapper = Enzyme.mount(<StepComponent/>);   
    wrapper.setState({
        error: null,
        isLoaded: true,
        items: [{"id":"d11b10ba-1cd8-48f8-93eb-454b716fd5a0","stepNumber":"2","versionContent":[{"title":"Request A Delivery","body":"Once you’re ready for your first delivery, all it takes is a click to get your shipment on the way.","effectiveDate":"2019-05-04T03:04:05.000Z"},{"title":"We Deliver","body":"Once you’re ready for your first delivery, all it takes is a click to get your shipment on the way.","effectiveDate":"2019-04-04T05:04:05.000Z"}]},{"id":"dce2554e-00dc-45c1-99c1-93a554d7eba7","stepNumber":"4","versionContent":[{"title":"Request Another Delivery","body":"Get your next gaming fix by updating your profile then initiating your next shipment.","effectiveDate":"2019-05-20T03:04:05.000Z"}]},{"id":"422e6b50-9c5a-43d5-90cb-839f4678cb75","stepNumber":"3","versionContent":[{"title":"Keep What You Like","body":"Tell us “no” by returning any unwanted products in the enclosed packaging.","effectiveDate":"2019-04-04T03:04:05.000Z"},{"title":"Keep What You Want","body":"Tell us “no thanks” by returning any unwanted products in the enclosed packaging.","effectiveDate":"2019-04-04T05:04:05.000Z"},{"title":"Keep Everything","body":"Tell us “no thanks” by returning any unwanted products in the enclosed packaging.","effectiveDate":"2019-02-04T08:04:05.000Z"}]},{"id":"d9a439d0-8dbf-4bab-8e08-626f8194a075","stepNumber":"1","versionContent":[{"title":"Fill Out a Profile","body":"We only want you to get games and gear that you’ll love, so we’ll ask for your preferences up front.","effectiveDate":"2019-05-20T03:04:05.000Z"}]}]
    });
    expect(toJson(wrapper)).toMatchSnapshot();
});