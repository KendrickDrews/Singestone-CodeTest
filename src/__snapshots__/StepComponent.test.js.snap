// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error page on error 1`] = `
<StepComponent>
  <div>
    Error: 
    test error
  </div>
</StepComponent>
`;

exports[`renders loading page on init 1`] = `
<StepComponent>
  <div>
    Loading...
  </div>
</StepComponent>
`;

exports[`renders steps with mock step data 1`] = `
<StepComponent>
  <div
    className="step-row"
  >
    <div
      className="step-container"
      key="1"
    >
      <p
        className="step-stepNumber"
      >
        0
        1
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        FILL OUT A PROFILE
      </p>
      <p
        className="step-body"
      >
        We only want you to get games and gear that you’ll love, so we’ll ask for your preferences up front.
      </p>
    </div>
    <div
      className="step-container"
      key="2"
    >
      <p
        className="step-stepNumber"
      >
        0
        2
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        REQUEST A DELIVERY
      </p>
      <p
        className="step-body"
      >
        Once you’re ready for your first delivery, all it takes is a click to get your shipment on the way.
      </p>
    </div>
    <div
      className="step-container"
      key="3"
    >
      <p
        className="step-stepNumber"
      >
        0
        3
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        KEEP WHAT YOU WANT
      </p>
      <p
        className="step-body"
      >
        Tell us “no thanks” by returning any unwanted products in the enclosed packaging.
      </p>
    </div>
    <div
      className="step-container"
      key="4"
    >
      <p
        className="step-stepNumber"
      >
        0
        4
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        REQUEST ANOTHER DELIVERY
      </p>
      <p
        className="step-body"
      >
        Get your next gaming fix by updating your profile then initiating your next shipment.
      </p>
    </div>
  </div>
</StepComponent>
`;

exports[`renders steps with step data 1`] = `
<StepComponent>
  <div
    className="step-row"
  >
    <div
      className="step-container"
      key="1"
    >
      <p
        className="step-stepNumber"
      >
        0
        1
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        FILL OUT A PROFILE
      </p>
      <p
        className="step-body"
      >
        We only want you to get games and gear that you’ll love, so we’ll ask for your preferences up front.
      </p>
    </div>
    <div
      className="step-container"
      key="2"
    >
      <p
        className="step-stepNumber"
      >
        0
        2
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        REQUEST A DELIVERY
      </p>
      <p
        className="step-body"
      >
        Once you’re ready for your first delivery, all it takes is a click to get your shipment on the way.
      </p>
    </div>
    <div
      className="step-container"
      key="3"
    >
      <p
        className="step-stepNumber"
      >
        0
        3
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        KEEP WHAT YOU WANT
      </p>
      <p
        className="step-body"
      >
        Tell us “no thanks” by returning any unwanted products in the enclosed packaging.
      </p>
    </div>
    <div
      className="step-container"
      key="4"
    >
      <p
        className="step-stepNumber"
      >
        0
        4
      </p>
      <hr
        className="step-separator"
      />
      <p
        className="step-title"
      >
        REQUEST ANOTHER DELIVERY
      </p>
      <p
        className="step-body"
      >
        Get your next gaming fix by updating your profile then initiating your next shipment.
      </p>
    </div>
  </div>
</StepComponent>
`;
